### PR TITLE
Eval hook

### PR DESCRIPTION
**Pull request by dreamerlin**

### CODE REVIEW

**Comment by dreamerlin** [comment]
It can not be merged util https://github.com/open-mmlab/mmcv/pull/695 is merged

**Comment by dreamerlin** [comment]
A related discuss, https://github.com/open-mmlab/mmaction2/pull/395#discussion_r548373375. Maybe we need to create rather than symlink

**Comment by ZwwWayne** [comment]
> A related discuss, [open-mmlab/mmaction2#395 (comment)](https://github.com/open-mmlab/mmaction2/pull/395#discussion_r548373375). Maybe we need to create rather than symlink

How about adding an option for such a case? By default, we create a symlink, but in some cases we can cp a new ckpt. And we should also ensure that only one symlink is created over the whole training time.

**Comment by dreamerlin** [comment]
> > A related discuss, [open-mmlab/mmaction2#395 (comment)](https://github.com/open-mmlab/mmaction2/pull/395#discussion_r548373375). Maybe we need to create rather than symlink
> 
> How about adding an option for such a case? By default, we create a symlink, but in some cases we can cp a new ckpt. And we should also ensure that only one symlink is created over the whole training time.

Now I think [open-mmlab/mmaction2#395 ](https://github.com/open-mmlab/mmaction2/pull/395) is a good proposal which cp the ckpt. Since mmcv hook has `max_keep_ckpts`, which may delete the original ckpt making the symlink for best invalid.

File: mmcv/runner/hooks/eval.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+
+        if interval <= 0:
+            raise ValueError(f'interval must be positive, but got {interval}')
+
+        assert isinstance(by_epoch, bool)
+
+        if start is not None and start < 0:
+            warnings.warn(
+                f'The evaluation start epoch {start} is smaller than 0, '
+                f'use 0 instead', UserWarning)
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/eval.py:75]: 'f' is unnecessary here.

File: mmcv/runner/hooks/eval.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+            start = 0
+        self.dataloader = dataloader
+        self.interval = interval
+        self.start = start
+        self.by_epoch = by_epoch
+
+        assert isinstance(save_best, str) or save_best is None
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/eval.py:82]: Add assersion message.

**Comment by ZwwWayne** [comment]
> > > A related discuss, [open-mmlab/mmaction2#395 (comment)](https://github.com/open-mmlab/mmaction2/pull/395#discussion_r548373375). Maybe we need to create rather than symlink
> > 
> > 
> > How about adding an option for such a case? By default, we create a symlink, but in some cases we can cp a new ckpt. And we should also ensure that only one symlink is created over the whole training time.
> 
> Now I think [open-mmlab/mmaction2#395 ](https://github.com/open-mmlab/mmaction2/pull/395) is a good proposal which cp the ckpt. Since mmcv hook has `max_keep_ckpts`, which may delete the original ckpt making the symlink for best invalid.

I am OK with that.

**Comment by ZwwWayne** [comment]
Please enrich the PR message for this PR as this is a big refactoring for downstream tasks to allow more discussion and for future reference, including its intention, modification, and consequences.

**Comment by ZwwWayne** [comment]
1. CI failed.
2. May add the contents/usages in PR messages into the documentation

**Comment by dreamerlin** [comment]
> 1. CI failed.
> 2. May add the contents/usages in PR messages into the documentation

1. CI failed due to `mmcv.engine`
2. Detailed PR message is updated

**Comment by ZwwWayne** [comment]
1. May add `multi_gpu_test` and `single_gpu_test` in this PR to avoid the dependency on entry point refactoring.
2. Evalhook needs to support synchronizing BN buffer of rank 0 models, see https://github.com/open-mmlab/mmdetection/pull/4582

File: mmcv/runner/hooks/eval.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        if not isinstance(dataloader, DataLoader):
+            raise TypeError(f'dataloader must be a pytorch DataLoader, '
+                            f'but got {type(dataloader)}')
+
+        if interval <= 0:
+            raise ValueError(f'interval must be positive, but got {interval}')
+
+        assert isinstance(by_epoch, bool), '``by_epoch`` should be a boolean'
+
+        if start is not None and start < 0:
```
Comment (hellock) [inline mmcv/runner/hooks/eval.py:76]: This is unexpected and we can raise an error.

File: mmcv/runner/hooks/eval.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+    Notes:
+        If new arguments are added for EvalHook, tools/test.py,
+        tools/eval_metric.py may be effected.
+
+    This hook will regularly perform evaluation in a given interval when
+    performing in non-distributed environment.
+
+    Args:
+        dataloader (DataLoader): A PyTorch dataloader.
```
Comment (hellock) [inline mmcv/runner/hooks/eval.py:23]: What dataloader?

File: mmcv/runner/hooks/eval.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+                warnings.warn('runner.meta is None. Creating a empty one.')
+                runner.meta = dict()
+            runner.meta.setdefault('hook_msgs', dict())
+
+    def before_train_iter(self, runner):
+        """Evaluate the model only at the start of training by iteration."""
+        if self.by_epoch:
+            return
+        if not self.initial_flag:
+            return
```
Comment (hellock) [inline mmcv/runner/hooks/eval.py:135]: `if self.by_epoch or not self.initial_flag`

File: mmcv/runner/hooks/eval.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+        """perform evaluation and save ckpt."""
+        if not self.evaluation_flag(runner):
+            return
+
+        results = single_gpu_test(runner.model, self.dataloader)
+        key_score = self.evaluate(runner, results)
+        if self.save_best:
+            self._save_ckpt(runner, key_score)
+
+    def evaluation_flag(self, runner):
```
Comment (hellock) [inline mmcv/runner/hooks/eval.py:170]: This can be a private method: `_should_evaluate`

File: mmcv/runner/hooks/eval.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        """Judge whether to perform_evaluation.
+
```
Comment (hellock) [inline mmcv/runner/hooks/eval.py:172]: Add a detailed description to explain the rule.

File: mmcv/runner/hooks/eval.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+            # No evaluation if start is larger than the current time.
+            return False
+        else:
+            # Evaluation only at epochs/iters 3, 5, 7...
+            # if start==3 and interval==2
+            if (current + 1 - self.start) % self.interval:
+                return False
+        return True
+
+    def _save_ckpt(self, runner, key_score):
```
Comment (hellock) [inline mmcv/runner/hooks/eval.py:197]: Add a docstring here.

File: mmcv/runner/hooks/eval.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+
+        assert isinstance(save_best, str) or save_best is None, \
+            '``save_best`` should be a str or None'
+        self.save_best = save_best
+        self.eval_kwargs = eval_kwargs
+        self.initial_flag = True
+
+        if self.save_best is not None:
+            self.best_ckpt_path = None
+            self._init_rule(rule, self.save_best)
```
Comment (hellock) [inline mmcv/runner/hooks/eval.py:94]: Add a comment here to indicate what attributes are set after this method.

File: mmcv/runner/hooks/eval.py [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+            best_ckpt_name = f'best_{self.key_indicator}_{current}.pth'
+            runner.save_checkpoint(
+                runner.work_dir, best_ckpt_name, create_symlink=False)
+            self.best_ckpt_path = osp.join(runner.work_dir, best_ckpt_name)
+            runner.meta['hook_msgs']['best_ckpt'] = self.best_ckpt_path
+            runner.logger.info(
+                f'Now best checkpoint is saved as {best_ckpt_name}.')
+            runner.logger.info(
+                f'Best {self.key_indicator} is {best_score:0.4f} '
+                f'at {cur_time} {cur_type}.')
```
Comment (hellock) [inline mmcv/runner/hooks/eval.py:223]: There are no comments in this method.

File: mmcv/runner/hooks/eval.py [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+
+        Args:
+            runner (:obj:`mmcv.Runner`): The underlined training runner.
+            results (list): Output results.
+        """
+        eval_res = self.dataloader.dataset.evaluate(
+            results, logger=runner.logger, **self.eval_kwargs)
+        for name, val in eval_res.items():
+            runner.log_buffer.output[name] = val
+        runner.log_buffer.ready = True
```
Comment (hellock) [inline mmcv/runner/hooks/eval.py:236]: Add a blank line after it.

File: mmcv/runner/hooks/eval.py [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+        if self.save_best is not None:
+            if self.key_indicator == 'auto':
+                # infer from eval_results
+                self._init_rule(self.rule, list(eval_res.keys())[0])
```
Comment (hellock) [inline mmcv/runner/hooks/eval.py:240]: Is the hardcoded `[0]` proper?

Comment (dreamerlin) [reply mmcv/runner/hooks/eval.py]: Proper. If ``save_best`` is ``auto``, the first key of the returned ``OrderedDict`` result will be used.

File: mmcv/runner/hooks/eval.py [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+            key_indicator (str | None): Key indicator to determine the
+                comparison rule.
+        """
+        if rule not in self.rule_map and rule is not None:
+            raise KeyError(f'rule must be greater, less or None, '
+                           f'but got {rule}.')
+
+        if rule is None:
+            if key_indicator != 'auto':
+                if any(key in key_indicator for key in self.greater_keys):
```
[the remark this answers is not in the document]
Comment (dreamerlin) [reply mmcv/runner/hooks/eval.py]: Add two if branch to check the rule whether in the xxx_keys

**Comment by dreamerlin** [comment]
> 1. May add `multi_gpu_test` and `single_gpu_test` in this PR to avoid the dependency on entry point refactoring.
> 2. Evalhook needs to support synchronizing BN buffer of rank 0 models, see [open-mmlab/mmdetection#4582](https://github.com/open-mmlab/mmdetection/pull/4582)

Done

File: mmcv/engine/test.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+        results.extend(result)
+
+        # use the first key as main key to calculate the batch size
+        batch_size = len(next(iter(data.values())))
+        for _ in range(batch_size):
+            prog_bar.update()
+    return results
+
+
+def multi_gpu_test(model, data_loader, tmpdir=None, gpu_collect=False):
```
Comment (ZwwWayne) [inline mmcv/engine/test.py:42]: This function, `collect_results_cpu`, and `collect_results_gpu` diverges from that in [mmclassification](https://github.com/open-mmlab/mmclassification/blob/master/mmcls/apis/test.py) which better handles some corner cases. May update them and involve @LXXXXR to have a look.

File: mmcv/engine/test.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+            ordered_results.extend(list(res))
+        # the dataloader may pad some samples
+        ordered_results = ordered_results[:size]
+        # remove tmp dir
+        shutil.rmtree(tmpdir)
+        return ordered_results
+
+
+def collect_results_gpu(result_part, size):
+    rank, world_size = get_dist_info()
```
Comment (ZwwWayne) [inline mmcv/engine/test.py:130]: Docstring

File: mmcv/engine/test.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+    # collect results from all ranks
+    if gpu_collect:
+        results = collect_results_gpu(results, len(dataset))
+    else:
+        results = collect_results_cpu(results, len(dataset), tmpdir)
+    return results
+
+
+def collect_results_cpu(result_part, size, tmpdir=None):
+    rank, world_size = get_dist_info()
```
Comment (ZwwWayne) [inline mmcv/engine/test.py:87]: docstring

File: mmcv/engine/test.py (hunk 14 above)
Comment (dreamerlin) [reply mmcv/engine/test.py:42]: @LXXXXR take a look

File: mmcv/engine/test.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+    dist.barrier()
+    # collect all parts
+    if rank != 0:
+        return None
+    else:
+        # load results of all parts from tmp dir
+        part_list = []
+        for i in range(world_size):
+            part_file = osp.join(tmpdir, f'part_{i}.pkl')
+            part_list.append(mmcv.load(part_file))
```
Comment (LXXXXR) [inline mmcv/engine/test.py]: Hi, it would be better if we only append `mmcv.load(part_file)` in `part_list` when it is not empty, otherwise an empty result on one gpu will cause the overall result to be empty when sorting the results. This could happen in extreme case where data is severely insufficient. Pls see [mmcls](https://github.com/open-mmlab/mmclassification/blob/bc50a7927bffb89e9ff7064e2ae9dafc51ae44b9/mmcls/apis/test.py#L149) for reference. Thx.

File: mmcv/engine/test.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+    part_recv_list = [
+        part_tensor.new_zeros(shape_max) for _ in range(world_size)
+    ]
+    # gather all result part
+    dist.all_gather(part_recv_list, part_send)
+
+    if rank == 0:
+        part_list = []
+        for recv, shape in zip(part_recv_list, shape_list):
+            part_list.append(
```
Comment (LXXXXR) [inline mmcv/engine/test.py]: Same here. It would be great if you could take a look: [mmcls](https://github.com/open-mmlab/mmclassification/blame/master/mmcls/apis/test.py#L186). Thx.

File: mmcv/engine/test.py (hunk 14 above)
Comment (LXXXXR) [reply mmcv/engine/test.py:42]: Looks great to me. Thx.

File: mmcv/runner/hooks/evaluation.py [hunk 19]
```diff
@@ -0,0 +1,374 @@
...
+            None, it will infer a reasonable rule. Keys such as 'acc', 'top'
+            .etc will be inferred by 'greater' rule. Keys contain 'loss' will
+            be inferred by 'less' rule. Options are 'greater', 'less', None.
+            Default: None.
+        **eval_kwargs: Evaluation arguments fed into the evaluate function of
+            the dataset.
+
+    Notes:
+        If new arguments are added for EvalHook, tools/test.py,
+        tools/eval_metric.py may be effected.
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/evaluation.py:49]: effected -> affected

File: mmcv/runner/hooks/evaluation.py [hunk 20]
```diff
@@ -0,0 +1,374 @@
...
+    """
+
+    # Since the key for determine greater an less is related to the downstream
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/evaluation.py:52]: an -> or?

File: mmcv/runner/hooks/evaluation.py [hunk 21]
```diff
@@ -0,0 +1,374 @@
...
+    # task, downstream repos may need to overwrite the following inner
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/evaluation.py:53]: tasks

File: mmcv/runner/hooks/evaluation.py [hunk 22]
```diff
@@ -0,0 +1,374 @@
...
+    # variable according to their tasks
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/evaluation.py:54]: according to their tasks -> accordingly.

File: mmcv/runner/hooks/evaluation.py [hunk 23]
```diff
@@ -0,0 +1,374 @@
...
+            raise ValueError(f'The evaluation start epoch {start} is smaller '
+                             f'than 0')
+
+        self.dataloader = dataloader
+        self.interval = interval
+        self.start = start
+        self.by_epoch = by_epoch
+
+        assert isinstance(save_best, str) or save_best is None, \
+            '``save_best`` should be a str or None'
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/evaluation.py:89]: 1. may use """" rather than ```` for save_best?
2. add information about what type is `save_best`.

File: mmcv/runner/hooks/evaluation.py [hunk 24]
```diff
@@ -0,0 +1,374 @@
...
+            save_best=save_best,
+            rule=rule,
+            **eval_kwargs)
+        self.broadcast_bn_buffer = broadcast_bn_buffer
+        self.tmpdir = tmpdir
+        self.gpu_collect = gpu_collect
+
+    def _do_evaluate(self, runner):
+        """perform evaluation and save ckpt."""
+
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/evaluation.py:342]: [Nit]: May clean the empty line

**Review by ZwwWayne — APPROVED** [review]

**Comment by ZwwWayne** [comment]
LGTM if the comments can all be resolved.

**Comment by dreamerlin** [comment]
> LGTM if the comments can all be resolved.

Done

**Comment by ZwwWayne** [comment]
See if @hellock have any comments.

**Comment by ZwwWayne** [comment]
1. Please resolve conflicts.
2. kindly ping @hellock 
3. May need MMAction2 and MMDetection to create a PR based on this PR to verify the correctness.

**Comment by LXXXXR** [comment]
Validation has been done in mmcls and the performance remaines as it used to be. Thank you.

**Comment by wHao-Wu** [comment]
Validation has been successfully done in mmdet3d with mmdet's PR: https://github.com/open-mmlab/mmdetection/pull/4806

**Review by congee524 — APPROVED** [review]
have tested EvalHook & DistEvalHook on MMAction2.

The related PR "Use mmcv eval hook in mmaction2" will be raised after the release of the next version of mmcv.

**Review by xvjiarui — APPROVED** [review]

**Comment by xvjiarui** [comment]
Validated in MMSeg. Thx!

**Comment by ZwwWayne** [comment]
This PR is merged as many downstream repos have verified its correctness. Thanks the efforts of all.
